File: serene/src/Serene.Web/Modules/Membership/Account/SignUp/SignUpForm.cs
namespace Serene.Membership;

[FormScript("Membership.SignUp")]
public class SignUpForm
{
    [Required(true), Placeholder("full name")]
    public string DisplayName { get; set; }
    [EmailAddressEditor, Required(true), Placeholder("email")]
    public string Email { get; set; }
    [EmailAddressEditor, Required(true), Placeholder("confirm email")]
    public string ConfirmEmail { get; set; }
    [PasswordEditor, Required(true), Placeholder("password")]
    public string Password { get; set; }
    [PasswordEditor, Required(true), Placeholder("confirm password")]
    public string ConfirmPassword { get; set; }
}

File: serene/src/Serene.Web/Modules/ServerTypes/Membership/SignUpForm.ts
import { EmailAddressEditor, initFormType, PasswordEditor, PrefixedContext, StringEditor } from "@serenity-is/corelib";

export interface SignUpForm {
    DisplayName: StringEditor;
    Email: EmailAddressEditor;
    ConfirmEmail: EmailAddressEditor;
    Password: PasswordEditor;
    ConfirmPassword: PasswordEditor;
}

export class SignUpForm extends PrefixedContext {
    static readonly formKey = 'Membership.SignUp';
    declare private static init: boolean;

    constructor(...args: ConstructorParameters<typeof PrefixedContext>) {
        super(...args);

        if (!SignUpForm.init) {
            SignUpForm.init = true;

            initFormType(SignUpForm, [
                'DisplayName', StringEditor,
                'Email', EmailAddressEditor,
                'ConfirmEmail', EmailAddressEditor,
                'Password', PasswordEditor,
                'ConfirmPassword', PasswordEditor
            ]);
        }
    }
}